### PR TITLE
Added sphinx-view-build-html to compile and view HTML output

### PR DESCRIPTION
**Pull request by jesse-michael-han**

### CODE REVIEW

File: sphinx-mode.el [hunk 1]
```diff
@@ -122,10 +123,49 @@ If BUFFER is not given use the `current-buffer'."
     (find-file (plist-get target :file))
     (goto-char (plist-get target :point))))
 
+(defun sphinx-compile-find-makefile-dir () ; adapted from rst-compile-find-conf
```
Comment (Fuco1) [inline sphinx-mode.el]: I think this entire function can be replaced by the built-in 

`(locate-dominating-file (buffer-file-name) "Makefile")`.

File: sphinx-mode.el [hunk 2]
```diff
@@ -122,10 +123,49 @@ If BUFFER is not given use the `current-buffer'."
...
+				      (file-name-directory dir))))))
+      (or (and dir (concat dir "")) nil))))
+
+;; TODO: replace this with sphinx-compile and sphinx-compile-and-view
+(defun sphinx-view-build-html ()
+  "Runs 'make html' in project root directory and then views compiled HTML of current source file."
+  (interactive)
+    (let
+      (
+       (buffer-name-base (file-name-base (buffer-name))) ; reads name of current bufer (presumably a .rst file), minus extension, into buffer-name-base
```
Comment (Fuco1) [inline sphinx-mode.el]: the comments probably aren't necessary.

Why are you using `buffer-name` instead of `buffer-file-name`?  People can rename buffers or use something like `uniquify` which would break this.

File: sphinx-mode.el [hunk 3]
```diff
@@ -122,10 +123,49 @@ If BUFFER is not given use the `current-buffer'."
...
+       (project-root-dir (sphinx-compile-find-makefile-dir)) ; finds parent directory containing makefile and reads into project-root-dir
+       )
+      (let
+	  (
+	   (file-rel-path (string-remove-prefix project-root-dir (file-name-directory buffer-file-name)))
```
Comment (Fuco1) [inline sphinx-mode.el]: There's a function `(file-relative-name FILENAME &optional DIRECTORY)` for this.

File: sphinx-mode.el [hunk 4]
```diff
@@ -122,10 +123,49 @@ If BUFFER is not given use the `current-buffer'."
...
+	   )
+        (shell-command (concat "make -C " project-root-dir " html"))
+	(browse-url (concat (concat (concat project-root-dir "_build/html/" file-rel-path) buffer-name-base) ".html"))
+	)
+    )
+  )
```
Comment (Fuco1) [inline sphinx-mode.el]: Please align the closing parens on one line as per lisp customs.

File: sphinx-mode.el [hunk 5]
```diff
@@ -122,10 +123,49 @@ If BUFFER is not given use the `current-buffer'."
...
+
 (defvar sphinx-mode-map
   (let ((map (make-sparse-keymap)))
     (define-key map (kbd "M-'") 'sphinx-goto-ref)
     (define-key map (kbd "C-c TAB") 'sphinx-insert-ref)
+    (define-key map (kbd "C-c C-c C-v") 'sphinx-view-build-html)
```
Comment (Fuco1) [inline sphinx-mode.el]: I think `C-c C-x` is a safer prefix to bind.

**Review by Fuco1 — CHANGES_REQUESTED** [review]

**Comment by Fuco1** [comment]
Hey, nice idea!

There is a bunch of code where you could use built-in functions instead of implementing it yourself.  

What about the TODO there?

**Comment by jesse-michael-han** [comment]
Thanks for the detailed suggestions! I've implemented them and also the TODO comment. Maybe there are still some style issues; if so, let me know and I'll fix them.

File: sphinx-mode.el [hunk 6]
```diff
@@ -123,49 +122,54 @@ If BUFFER is not given use the `current-buffer'."
     (find-file (plist-get target :file))
     (goto-char (plist-get target :point))))
 
-(defun sphinx-compile-find-makefile-dir () ; adapted from rst-compile-find-conf
-  "Look for the Sphinx Makefile in parents of the current path and returns the directory which contains it as a string."
+(defun sphinx-compile ()
+  "Runs 'make' in project root directory, prompting user for target format. Run 'make help' to see list of target formats."
```
Comment (Fuco1) [inline sphinx-mode.el]: Use `Run` instead of `Runs` (imperative over indicative), that's the style Emacs follows.

Also try to keep the first line under 80 chars and the rest in a separate paragraph.  I think the second line can be on a new line (separated by space).

You can run `checkdoc` command in Emacs and it will provide some suggestions.

File: sphinx-mode.el [hunk 7]
```diff
@@ -123,49 +122,54 @@ If BUFFER is not given use the `current-buffer'."
...
-	(setq prevdir dir)
-        (setq dir (expand-file-name (file-name-directory
-                                     (directory-file-name
-				      (file-name-directory dir))))))
-      (or (and dir (concat dir "")) nil))))
-
-;; TODO: replace this with sphinx-compile and sphinx-compile-and-view
-(defun sphinx-view-build-html ()
-  "Runs 'make html' in project root directory and then views compiled HTML of current source file."
+    (let
```
Comment (Fuco1) [inline sphinx-mode.el]: The default lisp style is to keep all the closing parens on the same line.  The one opening the `let` is also together, so 

```
(let ((foo bar)
      (... ...)))
```

File: sphinx-mode.el [hunk 8]
```diff
@@ -123,49 +122,54 @@ If BUFFER is not given use the `current-buffer'."
...
+	(xdg-open (expand-file-name
+		   (concat project-root-dir "_build/" target-format "/" (last (file-name-all-completions (file-name-sans-extension file-rel-path)))))))))))
 
 (defvar sphinx-mode-map
   (let ((map (make-sparse-keymap)))
     (define-key map (kbd "M-'") 'sphinx-goto-ref)
     (define-key map (kbd "C-c TAB") 'sphinx-insert-ref)
-    (define-key map (kbd "C-c C-c C-v") 'sphinx-view-build-html)
+    (define-key map (kbd "C-c C-x C-c") 'sphinx-compile)
+    (define-key map (kbd "C-c C-x C-v") 'sphinx-compile-and-view)
```
Comment (Fuco1) [inline sphinx-mode.el]: Could we find a way to reuse `sphinx-compile` in `sphinx-compile-and-view`.  There seems to be a lot of similar code.

**Review by Fuco1 — CHANGES_REQUESTED** [review]

**Comment by Fuco1** [comment]
Thanks Jesse, nice work!  There's a couple more ideas/suggestions.  If you have some questions feel free to ask!

File: sphinx-mode.el [hunk 9]
```diff
@@ -122,10 +122,54 @@ If BUFFER is not given use the `current-buffer'."
...
+      (
+       (buffer-name-base (file-name-base (buffer-file-name)))
+       (project-root-dir (locate-dominating-file buffer-file-name "Makefile"))
+       )
+      (let
+	  (
+	   (file-rel-path (file-relative-name buffer-file-name project-root-dir))
+	   )
+        (shell-command (concat "make -C " project-root-dir " html"))
+  (interactive)
```
Comment (Fuco1) [inline sphinx-mode.el]: It seems you duplicated this `(interactive)` form here probaby by accident when refactoring the code.

File: sphinx-mode.el [hunk 10]
```diff
@@ -109,60 +109,38 @@ If BUFFER is not given use the `current-buffer'."
 
 ;; TODO: add better default
 (defun sphinx-goto-ref (ref)
-  (interactive
+  (interactive)
```
Comment (Fuco1) [inline sphinx-mode.el]: Why this change?  The let block has to be inside interactive for this to work.

**Review by Fuco1 — CHANGES_REQUESTED** [review]

File: sphinx-mode.el (hunk 10 above)
Comment (Fuco1) [reply sphinx-mode.el]: Oh you're fast :)

Comment (jesse-michael-han) [reply sphinx-mode.el]: Oops! Accidentally switched around some parentheses, just fixed.

File: sphinx-mode.el [hunk 11]
```diff
@@ -122,10 +122,32 @@ If BUFFER is not given use the `current-buffer'."
...
+
+To see list of target formats, run 'make help' in a shell."
+  (interactive)
+    (let ((buffer-name-base (file-name-base (buffer-file-name)))
+       (project-root-dir (locate-dominating-file buffer-file-name "Makefile"))
+       (target-format (read-string "Make (default html): " nil nil "html" nil)))
+      (let* ((file-rel-path (file-relative-name buffer-file-name project-root-dir))
+(build-directory (expand-file-name (concat project-root-dir "_build/" target-format "/"))))
+	(shell-command (concat "make -C " project-root-dir " " target-format))
+	(concat build-directory (car (file-name-all-completions (file-name-base file-rel-path) build-directory))))))
```
Comment (Fuco1) [inline sphinx-mode.el:138]: Is the purpose of `file-name-all-completions` to turn `foo.rst` into `foo.html`?  That's pretty clever :)

**Comment by Fuco1** [comment]
Really cool!  Unless you have something more to add I think we can merge this :)

**Review by Fuco1 — APPROVED** [review]

**Comment by Fuco1** [comment]
Thanks!

**Comment by Fuco1** [comment]
Oh, now that I think about it, we should probably add a mention of this feature in the readme.  Do you mind updating that? (new PR is OK).
